integrators/symmetric_linear_multistep_integrator_body.hpp

```diff
@@ -0,0 +1,500 @@
+  Time const& step = down_cast_instance->step;
+
+  auto& previous_steps = down_cast_instance->previous_steps;
+
+  if (previous_steps.size() < order_ - 1) {
+    StartupSolve(t_final, *down_cast_instance);
+  }
+
+  // Argument checks.
+  int const dimension = previous_steps.back().displacements.size();
+  CHECK_LT(Time(), step);
+
+  // Time step.
+  Time const& h = step;
+  // Current time.
+  DoublePrecision<Instant> t = previous_steps.back().time;
+  // Order.
+  int const k = order_;
+
+  std::vector<DoublePrecision<Position>> Σj_minus_ɑj_qj(dimension);
+  std::vector<Acceleration> Σj_βj_numerator_aj(dimension);
+  while (h <= (t_final - t.value) - t.error) {
+    // We take advantage of the symmetry to iterate on the list of previous
+    // steps from both ends.
+    auto front_it = previous_steps.begin();
+    auto back_it = previous_steps.rbegin();
+
+    // This block corresponds to j = 0.  We must not pair it with j = k.
+    {
+      std::vector<Displacement> const& qj = front_it->displacements;
+      std::vector<Acceleration> const& aj = front_it->accelerations;
+      double const ɑj = ɑ_[0];
+      double const βj_numerator = β_numerator_[0];
+      for (int d = 0; d < dimension; ++d) {
+        Σj_minus_ɑj_qj[d] = -ɑj * qj[d];
+        Σj_βj_numerator_aj[d] = βj_numerator * aj[d];
+      }
+      ++front_it;
+    }
+    // The generic value of j, paired with k - j.
+    for (int j = 1; j < k / 2; ++j) {
+      std::vector<Displacement> const& qj = front_it->displacements;
+      std::vector<Displacement> const& qk_minus_j = back_it->displacements;
+      std::vector<Acceleration> const& aj = front_it->accelerations;
+      std::vector<Acceleration> const& ak_minus_j = back_it->accelerations;
+      double const ɑj = ɑ_[j];
+      double const βj_numerator = β_numerator_[j];
+      for (int d = 0; d < dimension; ++d) {
+        Σj_minus_ɑj_qj[d].Increment(-ɑj * (qj[d] + qk_minus_j[d]));
+        Σj_βj_numerator_aj[d] += βj_numerator * (aj[d] + ak_minus_j[d]);
+      }
+      ++front_it;
+      ++back_it;
+    }
+    // This block corresponds to j = k / 2.  We must not pair it with j = k / 2.
+    {
+      std::vector<Displacement> const& qj = front_it->displacements;
+      std::vector<Acceleration> const& aj = front_it->accelerations;
+      double const ɑj = ɑ_[k / 2];
+      double const βj_numerator = β_numerator_[k / 2];
+      for (int d = 0; d < dimension; ++d) {
+        Σj_minus_ɑj_qj[d].Increment(-ɑj * qj[d]);
+        Σj_βj_numerator_aj[d] += βj_numerator * aj[d];
+      }
+    }
+
+    // Create a new step in the instance.
+    t.Increment(h);
+    previous_steps.pop_front();
+    previous_steps.emplace_back();
+    Step& current_step = previous_steps.back();
+    current_step.time = t;
+    current_step.accelerations.resize(dimension);
+
+    // Fill the new step.  We skip the division by ɑk as it is equal to 1.0.
+    double const ɑk = ɑ_[0];
+    CHECK_EQ(ɑk, 1.0);
```

Comment:
Do we have to do this in a loop? it seems like it should be done at construction.